### PR TITLE
SAAS-136 Add STT permissions tests

### PR DESCRIPTION
**Pull request by artemgavrilov**
- https://jira.percona.com/browse/SAAS-136

FB: https://github.com/Percona-Lab/pmm-submodules/pull/876

Those tests check access permissions to endpoints required for STT data visualization.

### CODE REVIEW

File: server/auth_test.go [hunk 1]
```diff
@@ -229,3 +219,147 @@ func TestSwagger(t *testing.T) {
...
+	})
+	require.NoError(t, err)
+
+	req, err := http.NewRequest(http.MethodPatch, u.String(), bytes.NewReader(data))
+	require.NoError(t, err)
+
+	req.Header.Set("Content-Type", "application/json; charset=utf-8")
+	resp, b := doRequest(t, http.DefaultClient, req)
+
+	require.Equal(t, http.StatusOK, resp.StatusCode, fmt.Sprintf("failed to set role for user, status code: %d, response: %s", resp.StatusCode, b))
```
Comment (BupycHuk) [inline server/auth_test.go:364]: let's use `require.Equalf`

File: server/auth_test.go [hunk 2]
```diff
@@ -229,3 +219,147 @@ func TestSwagger(t *testing.T) {
...
+	})
+	require.NoError(t, err)
+
+	req, err := http.NewRequest(http.MethodPost, u.String(), bytes.NewReader(data))
+	require.NoError(t, err)
+
+	req.Header.Set("Content-Type", "application/json; charset=utf-8")
+
+	resp, b := doRequest(t, http.DefaultClient, req)
+	require.Equal(t, http.StatusOK, resp.StatusCode, fmt.Sprintf("failed to create user, status code: %d, response: %s", resp.StatusCode, b))
```
Comment (BupycHuk) [inline server/auth_test.go:338]: `require.Equalf`

File: .golangci.yml [hunk 3]
```diff
@@ -24,6 +24,11 @@ linters-settings:
 
 linters:
   enable-all: true
+  disable:
```
Comment (BupycHuk) [inline .golangci.yml:27]: :+1:

File: server/auth_test.go [hunk 4]
```diff
@@ -229,3 +219,147 @@ func TestSwagger(t *testing.T) {
...
+	}
+
+	for _, test := range tests {
+		test := test
+		t.Run("get settings/"+test.name, func(t *testing.T) {
+			// make a BaseURL without authentication
+			u, err := url.Parse(pmmapitests.BaseURL.String())
+			require.NoError(t, err)
+			u.User = url.UserPassword(test.login, test.login)
+			u.Path = "/v1/Settings/Get"
```
Comment (BupycHuk) [inline server/auth_test.go:260]: it looks a little bit different from other tests

File: server/auth_test.go [hunk 5]
```diff
@@ -229,3 +219,147 @@ func TestSwagger(t *testing.T) {
...
+	}{
+		{name: "default", login: none, statusCode: http.StatusUnauthorized},
+		{name: "viewer", login: viewer, statusCode: http.StatusUnauthorized},
+		{name: "editor", login: editor, statusCode: http.StatusUnauthorized},
+		{name: "admin", login: admin, statusCode: http.StatusOK},
+	}
+
+	for _, test := range tests {
+		test := test
+		t.Run("get settings/"+test.name, func(t *testing.T) {
```
Comment (BupycHuk) [inline server/auth_test.go:255]: looks like we have the same code in both tests except path, maybe we can extract it to an array and just loop over it?

File: server/auth_test.go (hunk 1 above)
Comment (artemgavrilov) [reply server/auth_test.go]: Fixed

File: server/auth_test.go (hunk 2 above)
Comment (artemgavrilov) [reply server/auth_test.go]: Fixed

File: server/auth_test.go (hunk 5 above)
Comment (artemgavrilov) [reply server/auth_test.go]: Done

File: server/auth_test.go [hunk 6]
```diff
@@ -229,3 +219,142 @@ func TestSwagger(t *testing.T) {
...
+	defer deleteUser(t, adminID)
+
+	tests := []struct {
+		name       string
+		url        string
+		method     string
+		login      string
+		statusCode int
+	}{
+		{name: "settings-default", url: "/v1/Settings/Get", method: http.MethodPost, login: none, statusCode: http.StatusUnauthorized},
```
Comment (BupycHuk) [inline server/auth_test.go:249]: why not just create another array and make for in for?

Comment (artemgavrilov) [reply server/auth_test.go:249]: To make it visually simple, one line - one test case.

Comment (AlekSi) [reply server/auth_test.go:249]: Do you guys really prefer `http.MethodPost` and `http.StatusUnauthorized` to `"POST"` and `401`? Uh!

File: server/auth_test.go [hunk 7]
```diff
@@ -229,3 +219,142 @@ func TestSwagger(t *testing.T) {
...
+
+func deleteUser(t *testing.T, userID int) {
+	u, err := url.Parse(pmmapitests.BaseURL.String())
+	require.NoError(t, err)
+	u.Path = "/graph/api/admin/users/" + strconv.Itoa(userID)
+
+	req, err := http.NewRequest(http.MethodDelete, u.String(), nil)
+	require.NoError(t, err)
+
+	req.Header.Set("Content-Type", "application/json; charset=utf-8")
```
Comment (AlekSi) [inline server/auth_test.go:307]: Looks a bit weird – do we really have to set Content-Type for bodiless DELETE request?

File: server/auth_test.go [hunk 8]
```diff
@@ -229,3 +219,142 @@ func TestSwagger(t *testing.T) {
...
+
+	resp, b := doRequest(t, http.DefaultClient, req)
+	require.Equalf(t, http.StatusOK, resp.StatusCode, "failed to delete user, status code: %d, response: %s", resp.StatusCode, b)
```
Comment (AlekSi) [inline server/auth_test.go:310]: resp.StatusCode will be logged anyway

File: server/auth_test.go (hunk 6 above)
Comment (artemgavrilov) [reply server/auth_test.go:249]: Well, yes. I always prefer predefined constants to string literals. I always thought that it is a common approach. 🤷‍♂️

File: server/auth_test.go (hunk 7 above)
Comment (artemgavrilov) [reply server/auth_test.go]: Removed

File: server/auth_test.go (hunk 6 above)
Comment (AlekSi) [reply server/auth_test.go:249]: Will you use `const post = "POST"` and `const code401 = 401`? I think not. And that's almost the same thing.

Comment (artemgavrilov) [reply server/auth_test.go:249]: But I will 😐. I mean, if I need some value multiple times I usually put in constant.

Ok, I've got you point, will change.

Comment (AlekSi) [reply server/auth_test.go:249]: >  if I need some value multiple times I usually put in constant

String and number literals a constants too. We will use a named constant if name is better than the value, or if type is different, or if value is shared and should be updated in a single place – but any of that is not the case there.

Comment (AlekSi) [reply server/auth_test.go]: To put it in another way: you don't define `const n1 = 1` even if you use the value `1` several times :)

Comment (artemgavrilov) [reply server/auth_test.go]: I eliminate typos and able easily navigate with IDE through such constants. 
Seems like "holy war" topic)

**Pull request merged**